public/app/features/alerting/unified/components/rule-editor/useAlertRuleSuggestions.tsx

```diff
@@ -32,34 +32,61 @@ export function useAlertRuleSuggestions(rulesSourceName: string) {
     }
   }, [features?.rulerConfig, fetchRulerRules]);
 
-  const namespaceGroups = useMemo(() => {
+  const labels = useMemo(() => {
     if (isPrometheusRulesLoading || isRulerRulesLoading) {
-      return new Map<string, string[]>();
+      return new Map<string, Set<string>>();
     }
 
     if (prometheusRulesPrimary) {
-      return promNamespacesToNamespaceGroups(promNamespaces);
+      return promNamespacesToLabels(promNamespaces);
     }
 
-    return rulerRulesToNamespaceGroups(rulerRules);
+    return rulerRulesToLabels(rulerRules);
   }, [promNamespaces, rulerRules, isPrometheusRulesLoading, isRulerRulesLoading]);
 
-  const labels = useMemo(() => {
+  return { labels, isLoading: isPrometheusRulesLoading || isRulerRulesLoading || isFeaturesLoading };
+}
+
+export function useGetNameSpacesByDatasourceName(rulesSourceName: string) {
+  const { data: features, isLoading: isFeaturesLoading } = useDiscoverDsFeaturesQuery({ rulesSourceName });
+
+  // emptyRulerConfig is used to prevent from triggering  labels' useMemo all the time
+  // rulerRules = {} creates a new object and triggers useMemo to recalculate labels
+  const [fetchRulerRules, { data: rulerRules = emptyRulerConfig, isLoading: isRulerRulesLoading }] =
+    useLazyRulerRulesQuery();
+
+  const { data: promNamespaces = [], isLoading: isPrometheusRulesLoading } = usePrometheusRuleNamespacesQuery(
+    { ruleSourceName: rulesSourceName },
+    { skip: !prometheusRulesPrimary }
+  );
+
+  useEffect(() => {
+    if (features?.rulerConfig && !prometheusRulesPrimary) {
+      fetchRulerRules({ rulerConfig: features.rulerConfig });
+    }
+  }, [features?.rulerConfig, fetchRulerRules]);
+
+  const namespaceGroups = useMemo(() => {
     if (isPrometheusRulesLoading || isRulerRulesLoading) {
-      return new Map<string, Set<string>>();
+      return new Map<string, string[]>();
```

Comment:
Is `string[]` better than `Set<string>`? `Set<string>` allow us to avoid duplicates

Reply:
ops, I think this change is not correct